README.md
# Google Adsense Ads for Laravel 6.x, 7.x and 8.x

[![Latest Version][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]
[![Build Status][ico-travis]][link-travis]
[![StyleCI][ico-styleci]][link-styleci]
[![Scrutinizer][ico-scrutinizer]][link-scrutinizer]
[![Code Coverage][ico-coverage]][link-coverage]
[![License][ico-license]][link-license]

Package for easily including Google Adsense Ad units in [Laravel 6.x, 7.x and 8.x][link-laravel] and [Lumen][link-lumen]. For use with Laravel 5.x use original [Mastergalen/Adsense-Ads package][link-mastergalen-adsense].

## Installation

### 1 - Dependency

In your project root run

The first step is using [Composer][link-composer] to install the package and automatically update your `composer.json` file, you can do this by running:

```shell
composer require martinbutt/laravel-adsense
```

### 2 - Set up config file

Run `php artisan config:publish martinbutt/laravel-adsense`.

Edit the generated config file in `/config/adsense.php` to add your ad units

```php
return [
    'client_id' => 'YOUR_CLIENT_ID', //Your Adsense client ID e.g. ca-pub-9508939161510421
    'ads' => [
        'responsive' => [
            'ad_slot' => 1111111111,
            'ad_format' => 'fluid',
            'ad_full_width_responsive' => true,
            'ad_style' => 'display:inline-block'
        ],
        'rectangle' => [
            'ad_slot' => 2222222222,
            'ad_style' => 'display:inline-block;width:300px;height:250px',
            'ad_full_width_responsive' => false,
            'ad_format' => 'auto'
        ]
    ]
];
```

### 3 - Register the provider with Laravel

You need to update your application configuration in order to register the package so it can be loaded by Laravel, just update your `config/app.php` file adding the following code at the end of your `'providers'` section:

> `config/app.php`

```php
<?php

return [
    // ...
    'providers' => [
        MartinButt\Laravel\Adsense\Providers\AdsenseServiceProvider::class,
        // ...
    ],
    // ...
];
```

#### Lumen

Go to `bootstrap/app.php` file and add this line:

```php
<?php
// ...

$app = new Laravel\Lumen\Application(
    dirname(__DIR__)
);

// ...

$app->register(MartinButt\Laravel\Adsense\Providers\AdsenseServiceProvider::class);

// ...

return $app;
```

### 4 - Register the alias with Laravel

> Note: facades are not supported in Lumen.

You may get access to the Google Adsense Ads services using following facades:

- `MartinButt\Laravel\Adsense\Facades\AdsenseFacade`

You can setup a short-version aliases for these facades in your `config/app.php` file. For example:

```php
<?php

return [
    // ...
    'aliases' => [
        'Adsense' => MartinButt\Laravel\Adsense\Facades\AdsenseFacade::class,
        // ...
    ],
    // ...
];
```

### 5 - Configuration

#### Publish config

In your terminal type

```shell
php artisan vendor:publish
```

or

```shell
php artisan vendor:publish --provider="MartinButt\Laravel\Adsense\Providers\AdsenseServiceProvider"
```

> Lumen does not support this command, for it you should copy the file `src/resources/config/adsense.php` to `config/adsense.php` of your project.

In `adsense.php` configuration file you can determine the properties of the default values and some behaviors.

## Usage
Add `{!! Adsense::javascript() !!}` in your `<head>` tag.

To show ads, add `{!! Adsense::ads('ads_unit') !!}`, where `ads_unit` is one of your ads units in your config file (for example `{!! Adsense::ads('responsive') !!}`).

Use `{!! Adsense::ads('ads_unit') !!}` every time you want to show an ad.

## Changelog

Please see the [CHANGELOG.md][link-changelog] file for more information on what has changed recently.

## Credits

- [Martin Butt][link-author]
- [Crypto Technology srl][link-coauthor1]
- [Luca Bognolo][link-coauthor2]
- [All Contributors][link-contributors]

## License

The Google Adsense Ads is open-sourced software licensed under the [MIT license][link-mit-license].  
Please see the [LICENSE.md][link-license] file for more information.

[ico-version]: https://img.shields.io/packagist/v/martinbutt/laravel-adsense.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/martinbutt/laravel-adsense.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/martinbutt/laravel-adsense/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/359045604/shield?style=flat-square
[ico-scrutinizer]: https://scrutinizer-ci.com/g/martinbutt/laravel-adsense/badges/quality-score.png?b=master
[ico-coverage]: https://scrutinizer-ci.com/g/martinbutt/laravel-adsense/badges/coverage.png
[ico-license]: https://img.shields.io/packagist/l/martinbutt/laravel-adsense?style=flat-square

[link-packagist]: https://packagist.org/packages/martinbutt/laravel-adsense
[link-downloads]: https://packagist.org/packages/martinbutt/laravel-adsense
[link-travis]: https://travis-ci.com/martinbutt/laravel-adsense
[link-styleci]: https://styleci.io/repos/359045604
[link-scrutinizer]: https://scrutinizer-ci.com/g/martinbutt/laravel-adsense/?branch=master
[link-coverage]: https://scrutinizer-ci.com/g/martinbutt/laravel-adsense
[link-laravel]: https://laravel.com
[link-lumen]: https://lumen.laravel.com
[link-mastergalen-adsense]: https://github.com/Mastergalen/Adsense-Ads
[link-composer]: https://getcomposer.org
[link-license]: LICENSE.md
[link-changelog]: CHANGELOG.md
[link-author]: https://www.martinbutt.com
[link-coauthor1]: https://cryptotech.srl
[link-coauthor2]: https://bogny.eu
[link-contributors]: ../../contributors
[link-mit-license]: https://opensource.org/licenses/MIT


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3] - 2021-04-17
### Added
- Added Laravel 7.x support

## [1.2] - 2019-12-20
### Added
- Added new tests

### Fixed
- Fixed error on binding class
- Fixed some PhpDoc

## [1.1] - 2019-12-15
### Fixed
- Fixed missing of publishing views

## [1.0] - 2019-12-14
### Added
- Writed tests
- Added CD & CI automations

### Changed
- Support for Laravel 6.x

## [0.1.5] - 2017-09-01
### Added
- Support for Laravel 5.5

## [0.1.4] - 2017-02-04
### Added
- Support for Laravel 5.4

## [0.1.3] - 2016-11-30
### Added
- Support for Laravel 5.3

### Fixed
- Fix showing Ads Banner with fixed style

### Changed
- Use blade syntax for ad.blade.php

## [0.1.2] - 2016-01-03
### Added
- Support for Laravel 5.2

## [0.1.1] - 2015-07-01
### Removed
- Remove test function

## [0.1] - 2015-07-01
### Added
- Initial release

[Unreleased]: https://github.com/martinbutt/laravel-adsense/compare/v1.3...HEAD
[1.3]: https://github.com/martinbutt/laravel-adsense/compare/v1.2...v1.3
[1.2]: https://github.com/crypto-technology/laravel-adsense/compare/v1.1...v1.2
[1.1]: https://github.com/crypto-technology/laravel-adsense/compare/v1.0...v1.1
[1.0]: https://github.com/crypto-technology/laravel-adsense/compare/v0.1.5...v1.0
[0.1.5]: https://github.com/crypto-technology/laravel-adsense/compare/v0.1.4...v0.1.5
[0.1.4]: https://github.com/crypto-technology/laravel-adsense/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/crypto-technology/laravel-adsense/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/crypto-technology/laravel-adsense/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/crypto-technology/laravel-adsense/compare/v0.1...v0.1.1
[0.1]: https://github.com/crypto-technology/laravel-adsense/releases/tag/v0.1


tests/TestCase.php
<?php

/**
 * Google Adsense Ads for Laravel.
 *
 * Package for easily including Google Adsense Ad units
 * in Laravel and Lumen.
 *
 * @developer Martin Butt <https://www.martinbutt.com/>
 *
 * @copyright Copyright (c) 2021 Martin Butt
 * @license   MIT
 *
 * Copyright (c) 2016 Galen Han
 * Copyright (c) 2019 Crypto Technology srl
 * Copyright (c) 2021 Martin Butt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

declare(strict_types=1);

namespace MartinButt\Laravel\Adsense\Tests;

use MartinButt\Laravel\Adsense\Facades\AdsenseFacade;
use MartinButt\Laravel\Adsense\Providers\AdsenseServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

/**
 * Class TestCase.
 */
abstract class TestCase extends Orchestra
{
    /**
     * {@inheritdoc}
     */
    protected function getPackageProviders($app)
    {
        return [
            AdsenseServiceProvider::class,
        ];
    }

    /**
     * {@inheritdoc}
     */
    protected function getPackageAliases($app)
    {
        return [
            'Adsense' => AdsenseFacade::class,
        ];
    }

    /**
     * {@inheritdoc}
     */
    protected function getEnvironmentSetUp($app)
    {
        $app['config']->set('view.paths', [__DIR__.'/stubs/views']);
        $app['config']->set('adsense.client_id', 'ca-pub-123');
    }
}


tests/AdsenseTest.php
<?php

/**
 * Google Adsense Ads for Laravel.
 *
 * Package for easily including Google Adsense Ad units
 * in Laravel and Lumen.
 *
 * @developer Martin Butt <https://www.martinbutt.com/>
 *
 * @copyright Copyright (c) 2021 Martin Butt
 * @license   MIT
 *
 * Copyright (c) 2016 Galen Han
 * Copyright (c) 2019 Crypto Technology srl
 * Copyright (c) 2021 Martin Butt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

declare(strict_types=1);

namespace MartinButt\Laravel\Adsense\Tests;

use MartinButt\Laravel\Adsense\Facades\AdsenseFacade;
use Illuminate\View\View;

/**
 * Class AdsenseTest.
 */
class AdsenseTest extends TestCase
{
    /** @var \MartinButt\Laravel\Adsense\AdsenseBuilder $adsense */
    protected $adsense;

    public function setUp(): void
    {
        parent::setUp();
        $this->adsense = $this->app->make('adsense');
    }

    public function testAdsenseInstance()
    {
        $this->assertInstanceOf(View::class, $this->adsense->javascript());
    }

    public function testAdsenseView()
    {
        $return = $this->app->view->make('adsense::javascript')->render();
        $this->assertSame('<script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-123" crossorigin="anonymous"></script>', $return);
    }

    public function testAdsenseFacade()
    {
        $this->assertIsObject(AdsenseFacade::javascript());
    }
}


src/AdsenseBuilder.php
<?php

/**
 * Google Adsense Ads for Laravel.
 *
 * Package for easily including Google Adsense Ad units
 * in Laravel and Lumen.
 *
 * @developer Martin Butt <https://www.martinbutt.com/>
 *
 * @copyright Copyright (c) 2021 Martin Butt
 * @license   MIT
 *
 * Copyright (c) 2016 Galen Han
 * Copyright (c) 2019 Crypto Technology srl
 * Copyright (c) 2021 Martin Butt
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

declare(strict_types=1);

namespace MartinButt\Laravel\Adsense;

class AdsenseBuilder
{
    public function ads($unit)
    {
        return view('adsense::ads')->with([
            'ad_client' => config('adsense.client_id'),
            'ad_style' => config("adsense.ads.$unit.ad_style", 'display:block;'),
            'ad_slot' => config("adsense.ads.$unit.ad_slot"),
            'ad_format' => config("adsense.ads.$unit.ad_format"),
            'ad_full_width_responsive' => config("adsense.ads.$unit.ad_full_width_responsive"),
        ]);
    }

    public function javascript()
    {
        return view('adsense::javascript', ['client_id' => config('adsense.client_id')]);
    }
}


src/Facades/AdsenseFacade.php
<?php

/**
 * Google Adsense Ads for Laravel.
 *
 * Package for easily including Google Adsense Ad units
 * in Laravel and Lumen.
 *
 * @developer Martin Butt <https://www.martinbutt.com/>
 *
 * @copyright Copyright (c) 2021 Martin Butt
 * @license   MIT
 *
 * Copyright (c) 2016 Galen Han
 * Copyright (c) 2019 Crypto Technology srl
 * Copyright (c) 2021 Martin Butt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

declare(strict_types=1);

namespace MartinButt\Laravel\Adsense\Facades;

use Illuminate\Support\Facades\Facade;

/**
 * Class AdsenseFacade.
 *
 * @method static \Illuminate\View\View|\Illuminate\Contracts\View\Factory ads(string $ads)
 * @method static \Illuminate\View\View|\Illuminate\Contracts\View\Factory javascript()
 *
 * @see \Illuminate\Support\Facades\Facade
 */
final class AdsenseFacade extends Facade
{
    /**
     * {@inheritdoc}
     */
    protected static function getFacadeAccessor()
    {
        return 'adsense';
    }
}


src/Providers/AdsenseServiceProvider.php
<?php

/**
 * Google Adsense Ads for Laravel.
 *
 * Package for easily including Google Adsense Ad units
 * in Laravel and Lumen.
 *
 * @developer Martin Butt <https://www.martinbutt.com/>
 *
 * @copyright Copyright (c) 2021 Martin Butt
 * @license   MIT
 *
 * Copyright (c) 2016 Galen Han
 * Copyright (c) 2019 Crypto Technology srl
 * Copyright (c) 2021 Martin Butt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

declare(strict_types=1);

namespace MartinButt\Laravel\Adsense\Providers;

use Illuminate\Support\ServiceProvider;
use Laravel\Lumen\Application as LumenApplication;
use MartinButt\Laravel\Adsense\AdsenseBuilder;

/**
 * Class AdsenseServiceProvider.
 *
 * @see \Illuminate\Support\ServiceProvider
 */
class AdsenseServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap any application services.
     */
    public function boot(): void
    {
        $this->loadViewsFrom(__DIR__.'/../resources/views', 'adsense');

        if ($this->app instanceof LumenApplication) {
            /* @scrutinizer ignore-call */ $this->app->configure('adsense');
        } else {
            // Publishing the configuration file.
            $this->publishes([
                $this->getConfigFile() => config_path('adsense.php'),
            ], 'config');
            // Publishing the views.
            $this->publishes([
                __DIR__.'/../resources/views' => resource_path('views/vendor/adsense'),
            ], 'views');
        }
    }

    /**
     * {@inheritdoc}
     */
    public function register(): void
    {
        $this->mergeConfigFrom(
            $this->getConfigFile(),
            'adsense'
        );

        $this->app->bind(AdsenseBuilder::class, function () {
            return new AdsenseBuilder();
        });

        $this->app->alias(AdsenseBuilder::class, 'adsense');
    }

    /**
     * {@inheritdoc}
     */
    public function provides(): array
    {
        return [
            AdsenseBuilder::class,
            'adsense',
        ];
    }

    /**
     * Return the path of configuration file.
     */
    protected function getConfigFile(): string
    {
        return __DIR__.'/../resources/config/adsense.php';
    }
}


src/resources/config/adsense.php
<?php

/**
 * Google Adsense Ads for Laravel.
 *
 * Package for easily including Google Adsense Ad units
 * in Laravel and Lumen.
 *
 * @developer Martin Butt <https://www.martinbutt.com/>
 *
 * @copyright Copyright (c) 2021 Martin Butt
 * @license   MIT
 *
 * Copyright (c) 2016 Galen Han
 * Copyright (c) 2019 Crypto Technology srl
 * Copyright (c) 2021 Martin Butt
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

declare(strict_types=1);

return [
    'client_id' => 'YOUR_CLIENT_ID', //Your Adsense client ID e.g. ca-pub-9508939161510421
    'ads' => [
        'responsive' => [
            'ad_slot' => 1111111111,
            'ad_format' => 'fluid',
            'ad_full_width_responsive' => true,
            'ad_style' => 'display:inline-block',
        ],
        'rectangle' => [
            'ad_slot' => 2222222222,
            'ad_style' => 'display:inline-block;width:300px;height:250px',
            'ad_full_width_responsive' => false,
            'ad_format' => 'auto',
        ],
    ],
];
